src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -891,21 +809,143 @@ EnumeratorResult createPatternBasedEnumeratorResult(ref ForEachEnumeratorInfo.Bu
                     return EnumeratorResult.Succeeded;
                 }
 
-                MethodSymbol getEnumeratorMethod = builder.GetEnumeratorMethod;
+                MethodSymbol getEnumeratorMethod = builder.GetEnumeratorInfo.Method;
                 diagnostics.Add(isAsync ? ErrorCode.ERR_BadGetAsyncEnumerator : ErrorCode.ERR_BadGetEnumerator, _syntax.Expression.Location, getEnumeratorMethod.ReturnType, getEnumeratorMethod);
                 return EnumeratorResult.FailedAndReported;
             }
+        }
+
+        private EnumeratorResult SatisfiesIEnumerableInterfaces(ref ForEachEnumeratorInfo.Builder builder, ref BoundExpression collectionExpr, bool isAsync, DiagnosticBag diagnostics, BoundExpression unwrappedCollectionExpr, TypeSymbol unwrappedCollectionExprType)
+        {
+            if (!AllInterfacesContainsIEnumerable(ref builder, unwrappedCollectionExprType, isAsync, diagnostics, out bool foundMultipleGenericIEnumerableInterfaces))
```

Comment:
>AllInterfacesContainsIEnumerable [](start = 17, length = 32)

Shouldn't we match an interface to itself here? Looks like this is a preexisting condition, perhaps it is worth following up later.